### PR TITLE
CUDA: Fix linkage of device functions when compiling for debug

### LINKED ISSUE #7159: CUDA regression: Mixing `debug` kwargs between called functions results in linkage error
The following is fine with Numba 0.53.1:

```python
from numba import cuda

@cuda.jit(device=True, debug=True)
def f2(x):
    return x + 1

@cuda.jit(device=True)
def f1(x, y):
    return x - f2(y)

@cuda.jit(debug=True)
def kernel(x, y):
    f1(x, y)

kernel[1, 1](1, 2)
```

However with current master, linking fails:

```
$ python debuglinkage.py 
/home/gmarkall/numbadev/numba/numba/cuda/compiler.py:865: NumbaPerformanceWarning: Grid size (1) < 2 * SM count (144) will likely result in GPU under utilization due to low occupancy.
  warn(NumbaPerformanceWarning(msg))
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 2221, in complete
    driver.cuLinkComplete(self.handle, byref(cubin), byref(size))
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 302, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 342, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [999] Call to cuLinkComplete results in CUDA_ERROR_UNKNOWN

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gmarkall/numbadev/issues/cuda-extension-bug/debuglinkage.py", line 19, in <module>
    kernel[1, 1](1, 2)
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 868, in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 1001, in call
    kernel = _dispatcher.Dispatcher._cuda_call(self, *args)
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 1009, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 1115, in compile
    kernel.bind()
  File "/home/gmarkall/numbadev/numba/numba/cuda/compiler.py", line 626, in bind
    self._codelibrary.get_cufunc()
  File "/home/gmarkall/numbadev/numba/numba/cuda/codegen.py", line 188, in get_cufunc
    cubin = self.get_cubin(cc=device.compute_capability)
  File "/home/gmarkall/numbadev/numba/numba/cuda/codegen.py", line 165, in get_cubin
    cubin_buf, size = linker.complete()
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudadrv/driver.py", line 2223, in complete
    raise LinkerError("%s\n%s" % (e, self.error_log))
numba.cuda.cudadrv.driver.LinkerError: [999] Call to cuLinkComplete results in CUDA_ERROR_UNKNOWN
error   : Undefined reference to '_ZN8__main__6f2$243Ex' in '<cudapy-ptx>'
```

### PR DESCRIPTION
**Pull request by gmarkall**
Fixes #7159.

### CODE REVIEW

**Comment by gmarkall** [comment]
@esc @stuartarchibald Is this possible for a CUDA buildfarm run to see if it turns up any issues on any of the other platforms please?

**Comment by stuartarchibald** [comment]
> @esc @stuartarchibald Is this possible for a CUDA buildfarm run to see if it turns up any issues on any of the other platforms please?

Buildfarm ID: `numba_smoketest_cuda_yaml_80`.

**Comment by sklam** [comment]
I can get the reproducer in the original issue to work If I turn off the `debug` option when `!llvm.dbg.cu` is not in the LLVM IR like below:

```diff
diff --git a/numba/cuda/codegen.py b/numba/cuda/codegen.py
index 70dcb10d6..5cfdaf6f1 100644
--- a/numba/cuda/codegen.py
+++ b/numba/cuda/codegen.py
@@ -125,7 +125,7 @@ class CUDACodeLibrary(serialize.ReduceMixin, CodeLibrary):
         else:
             # Otherwise, we compile all modules with NVVM at once because this
             # results in better optimization than separate compilation.
-            ptxes = [nvvm.llvm_to_ptx(irs, **options)]
+            ptxes = [nvvm.llvm_to_ptx(ir, **options) for ir in irs]
 
         # Sometimes the result from NVVM contains trailing whitespace and
         # nulls, which we strip so that the assembly dump looks a little
diff --git a/numba/cuda/cudadrv/nvvm.py b/numba/cuda/cudadrv/nvvm.py
index eb631b703..963022c5c 100644
--- a/numba/cuda/cudadrv/nvvm.py
+++ b/numba/cuda/cudadrv/nvvm.py
@@ -706,6 +706,9 @@ def llvm_to_ptx(llvmir, **opts):
     for mod in llvmir:
         mod = llvm_replace(mod)
         cu.add_module(mod.encode('utf8'))
+
+        if "!llvm.dbg.cu" not in mod:
+            opts.pop("debug", None)
     cu.lazy_add_module(libdevice.get())
 
     return cu.compile(**opts)
```

however, the test will still fail with:

```pytb
numba.cuda.cudadrv.driver.LinkerError: [999] Call to cuLinkComplete results in CUDA_ERROR_UNKNOWN
error   : Undefined reference to '_ZN5numba4cuda5tests6cudapy14test_debuginfo17TestCudaDebugInfo29_test_chained_device_function12$3clocals$3e11kernel$2422Eii' in '<cudapy-ptx>'
```

**Comment by gmarkall** [comment]
@esc @sklam could this have another buildfarm run please?

**Comment by sklam** [comment]
BFID: `numba_smoketest_cuda_yaml_81`

**Comment by sklam** [comment]
BF passed!

**Review by sklam — APPROVED** [review]
The PR looks good to me. 

This PR will disable debuginfo on NVVM version < 7.0 and it is acceptable because the debug support is not ideal in the old (2014) version. We shouldn't spend too much time fixing for the old toolchain given that the CUDA community is likely to be on the latest CTK with the newer NVVM, for which Numba generates better debuginfo now.

File: numba/cuda/codegen.py [hunk 1]
```diff
@@ -113,6 +115,17 @@ def _get_ptxes(self, cc=None):
...
         options['arch'] = arch
+        if not nvvm.NVVM().is_nvvm70:
+            # Avoid enabling debug for NVVM 3.4 as it has various issues. We
+            # need to warn the user that we're doing this if any of the
+            # functions that they're compiling have `debug=True` set, which we
+            # can determine by checking the NVVM options.
+            debug_callee = any(lib._nvvm_options.get('debug')
+                               for lib in self.linking_libraries)
+            if options.get('debug') or debug_callee:
+                msg = "debuginfo is not generated for CUDA versions < 11.2"
```
Comment (stuartarchibald) [inline numba/cuda/codegen.py:126]: From OOB discussion, this message is going to be updated to include the function name so it's easy for users to find the function(s) causing this warning.

File: numba/cuda/tests/cudapy/test_debuginfo.py [hunk 2]
```diff
@@ -129,6 +135,140 @@ def kernel(arr):
             if i < len(arr):
                 arr[i] = threadid()
 
+    def _test_chained_device_function(self, kernel_debug, f1_debug, f2_debug):
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_debuginfo.py:138]: There are a fair few warnings from the test suite anyway - after 0.54 is released I was planning to have a big clear up of the test suite and warnings and get it all to be "silent" - are you happy if I defer this to when I approach that task?

**Comment by gmarkall** [comment]
@stuartarchibald Thanks for the review - I believe comments are addressed / answered now.

File: numba/cuda/tests/cudapy/test_debuginfo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cuda/tests/cudapy/test_debuginfo.py:138]: Sounds good, thanks.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Comment by stuartarchibald** [comment]
NOTE: I checked out the patch and manually tested the warnings etc in local code on a CTK=11.0. Also checked that the tests are fine under CTK=11.0 and CTK=11.2.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cuda_yaml_82` ran successfully on 68c4e76.

**Pull request merged**